taichi/backends/vulkan/vulkan_api.cpp

```diff
@@ -400,6 +449,11 @@ void EmbeddedVulkanDevice::create_logical_device() {
   capability_.has_int16 = false;
   capability_.has_int64 = true;
   capability_.has_float64 = true;
+  if (params_.is_for_ui) {
+    device_features.samplerAnisotropy = VK_TRUE;
+    device_features.geometryShader = VK_TRUE;
```

Comment:
Don't enable geometry shaders. They are a legacy pipeline stage.

Reply:
Also why do we enable anisotropic filtering?

Reply:
We will need geometry shaders to generate face normals when the user are not providing them. And not every user's machine will have mesh shader support (my own laptop, for example..)

Reply:
Compute shaders are the way to go for those, or we can generate them on the CPU. To be honest, geometry shaders are garbage, never use them. In addtion lot of devices don't support geometry shaders (here is a list https://vulkan.gpuinfo.org/listdevicescoverage.php?feature=geometryShader&platform=all&option=not)

Reply:
Sampler anisotropy is not necessary. And i'm fine with generating normals elsewhere.

Reply:
Out of curiosity tho, why are geom shaders garbage lol

Reply:
They don't fit in the pipeline nicely. Vertex shaders and fragment shaders are on different parallelization scheme (per vertex vs. per fragment), geometry shaders add another stage in between and therefore requires the GPU to reschedule. Also since vertex shaders may output not certain amount of primitives, it's very hard for the GPU to decide how to parallelize them efficiently. On some devices geometry shaders run fine, on some else, they run terribly. The performance is therefore not consistent and usually worse than a pipeline without it.

Fun fact BTW, mesh shaders can also be slower than normal pipelines even if they do the same thing. They allow more flexibility, but it's hard to compete with fixed function hardware.

Reply:
Cool stuff.